### ISSUE TITLE
Enhancement: Panel - remove tabIndex from internal content containers

### ISSUE THREAD

**Issue by macandcheese** [Esri/calcite-components#1642]
### Description
Hoping to prevent that unstyled focus ring around the slotted panel content... either by tabbing to content in the pane, or inadvertently clicking on non-interactive elements (background, header). 

This adds an extra container to tab through to reach meaningful content with keyboard, and for moused users it's kind of an annoyance. My use case is using Panels that are not dismissible.

I talked to Arjav (cc @arjavbadjatiya) and doesn't seem like there's an accessibility reason to keep those default tabIndex in there. This would be separate from any kind of `setFocus()` feature.

cc @asangma 

### Acceptance Criteria 
Tabbing into or inadvertently clicking inert elements within the calcite-panel shouldn't focus the content container.

#### Which Component
Panel

**Comment by driskull** [comment]
@macandcheese @arjavbadjatiya The reason the panel is focusable is because it can often be scrollable and aXe was saying that scrollable areas should have tabIndex 0 so that keyboard users can navigate to them and scroll them.

If we want to ignore that rule we can remove it.

https://dequeuniversity.com/rules/axe/4.0/scrollable-region-focusable

**Comment by driskull** [comment]
If we want to keep this rule, we can add it to conventions doc.

**Comment by driskull** [comment]
The other option is to have some kind of logic to see if a panel is scrollable or not. I'm thinking that is maybe complicated though.

**Comment by arjavbadjatiya** [comment]
We should definitely continue allowing scrolling to be accessed with a keyboard. I'm not sure why the background needs a focus state to enable that. How are we doing the keyboard scrolling in a modal like this one without adding the focus state to the background? @rpanichakit9541 implemented this one if I'm not mistaken. There is a little bit of scrolling that you can use arrow keys for, provided you're focused somewhere within the panel (and it's not an element that needs arrow keys within it). This is can be accessed via org page > settings > security > information banner.

<img width="1440" alt="Screen Shot 2021-03-02 at 10 14 51 AM" src="https://user-images.githubusercontent.com/29237960/109695264-8ab25b00-7b40-11eb-9d9a-82b92272238e.png">

**Comment by macandcheese** [comment]
My interpretation from that link is that as long as there is something focusable *inside* the scrolling container, it should be fine - they just use an autofocusing input which is kind of a confusing example maybe?

**Comment by driskull** [comment]
Yes that's correct. Sometimes a panel doesn't have a focusable element inside of it but should still be scrollable.

**Comment by arjavbadjatiya** [comment]
Makes sense. Do we have a lot of instances like that though @driskull ? If there are no focusable elements within it, what's the content that would go below the fold and require scrolling? Copy?

**Comment by driskull** [comment]
I'm not sure on that one.

**Comment by macandcheese** [comment]
Just to clarify here - it's not a concern to remove this @arjavbadjatiya - @driskull, it also seems like dismissible was handled explicitly on purpose, just want to make sure removing the two tabIndex declarations doesn't break anything intentional - @asangma do you have any insight here?

**Comment by driskull** [comment]
@macandcheese I'm not sure. All I know is that aXe says that if a panel is scrollable it needs tabindex for a user to control that scrolling.

**Comment by driskull** [comment]
> what's the content that would go below the fold and require scrolling? Lots of scrolling text?

Yes, I'm guessing anything that isn't focusable. text, images, etc.

**Comment by ffaubry** [comment]
Hello.
I use calcite-panel inside flow in the shell.
![image](https://user-images.githubusercontent.com/3506166/161643191-2b09f546-6e3d-437a-ba19-5cce295895cb.png)

I don't understand why the calcite-panel receives the focus when I try to navigate to its content or out of it's content.
I understand that the list should receive the focus as it can scroll. But when going out of the panel (from the dismiss action), the panel is focused.
Nothing in the UI is making this obvious.
For the user, it looks like there is an issue with the navigation.

**Comment by macandcheese** [comment]
@FelFly @geospatialem this may be a good thing to take another look at - some way of determining when we need to have the focusable element, and for other use cases allow the focusable element inside panel to receive focus to avoid the above.

**Comment by driskull** [comment]
we might be able to use a resize observer here to set the tabIndex only if scrolling is present.

**Comment by driskull** [comment]
I have a PR open here: https://github.com/Esri/calcite-components/pull/4374

Just need to setup a test.

**Comment by macandcheese** [comment]
Is scrolling what we should key off of - or scrolling **without the existence of a focusable element** ? 

For example, does this scrolling example need the focus ring, as it has focusable elements within?

https://codepen.io/mac_and_cheese/pen/gOovvzw?editors=1000

**Comment by driskull** [comment]
I think it should key off of scrollable regions . https://dequeuniversity.com/rules/axe/4.0/scrollable-region-focusable

**Comment by macandcheese** [comment]
Wouldn't it be an example of "passed example 2": https://www.w3.org/WAI/standards-guidelines/act/rules/0ssw9k/proposed/#testcases because "This scrollable section element contains a link that is included in sequential focus navigation." ?

@geospatialem any insight or thoughts here? 

My interpretation is that if the panel content is scrollable, with no focusable element, it needs a tabIndex - however if there is a focusable element inside that scrolling section, it does not, as that focusable element satisfies the requirement.

**Comment by driskull** [comment]
Yeah, that might be too complicated to handle that use case.

**Comment by geospatialem** [comment]
@macandcheese Sorry for joining in late in this discussion... I think @driskull is on the right track with the scrollable regions with a resize observer.

A quote from the <a href="https://dequeuniversity.com/rules/axe/4.0/scrollable-region-focusable" target="_blank">Deque University site</a> Matt mentioned, re: making scrollable regions:

> "The following examples could fail if the browser intercepts the keyboard events for autocomplete. It is better to always put a tabindex of 0 on the scrollable region or a static element within the region."

There are a few different strategies that we could take, but the focusable scrollable region seems to cover a wider audience by using the region has a tabindex.

**Comment by macandcheese** [comment]
Few thoughts based on that…

Should the focus ring at least be the styled version so it appears less broken?

Currently it only appears when tabbing into the panel, not shift tabbing back through it. Is this intentional?

For cases where designers / developers know there is a focusable element inside the panel and do not want to interrupt the flow with an extra focus / tab step - how should we facilitate / allow this? 

See @ffaubry use case above, or the dev site icon search, where we manually remove the internal focus as we know something inside is focusable.

**Comment by driskull** [comment]
> Should the focus ring at least be the styled version so it appears less broken?

Yes.

> Currently it only appears when tabbing into the panel, not shift tabbing back through it. Is this intentional?

I'm not sure on that one. I wasn't seeing that but it should be consistent.

> For cases where designers / developers know there is a focusable element inside the panel and do not want to interrupt the flow with an extra focus / tab step - how should we facilitate / allow this?

I think the extra step is necessary.

**Comment by driskull** [comment]
I tried to implement the dynamic tabindex via a resizeObserver but that seems to be throwing an error in the `hasSlots` common test. Not sure what to do there. @jcfranco
